### DISCUSSION TITLE
Update automatically with OTA Server

### DISCUSSION THREAD

**Discussion by adnan6336** [Hieromon/AutoConnect#417] [How can I?] [answered]
Hi guys,
Really nice repository, very much helpful.
I am stuck at a point where I want to update automatically when ever there is new update.
Explanation:
->A new update(xyzv1.bin) under my updateserver.py.
->esp32 hit an API once every minute and in response it knows that whether there is any new update or not, if there is then it also gets update filename in response(i-e xyzv1.bin).
->if there is any new update? call an **update function** so that esp32 update itself.
I just want to know how to call that function in which I put the server url and filename and it update by itself, by this I can able to make a device that automatically update itself when ever there is new update available on server.
I have searched in the repository and have seen something like that but not sure how to implement that function.
Would be nice if somebody point me some direction or give a little code that I can call.
Thanks.

**Comment by svdrummer** [comment]
I do this. Once a day is more than enough. Once per min is overkill and will cause customer concern.
Depending on your program function, would it be okay for the system to stop and update?
I give my customers option to manually check for update.
Just some thoughts before you start..especially checking every minute..what if the network is down, your system will spend more time searching than running.

**Reply by adnan6336** [reply]
Hi svdrummer,
many thanks for your suggestion, yes you are correct! once a day would be enough and yes my system can tolerate of atleast 1 minute downtime.
basically I can give two options in customer app that is, manual update or auto update but for both cases I need to know the function to call for example.
```
void loop()
{
if(updateFlag==1)
 {
 updateSys('www.myserver.com',8000,'xyz.bin');
 }
}
```
your help would be appreciated.

**Comment by Hieromon** [comment, marked as answer]
@adnan6336 

Unfortunately, I didn't offer an [AutoConnectUpdate](https://hieromon.github.io/AutoConnect/apiupdate.html) class with the auto-updates you expect. The AutoConnectUpdate class is supposed to make update decisions manually and has a UI for that, and it is provided in the form of being incorporated into AutoConnectAux. However, there are some cases where it is not suitable, such as automatic update, which is your task. It is the [**UpdateClass**](https://github.com/espressif/arduino-esp32/blob/master/libraries/Update/src/Update.h) that serves as the OTA via the HTTP server provided by the ESP32 arduino core, whether it is an automatic update or a manual update. And the wrapper classes are [**HTTPUpdate**](https://github.com/espressif/arduino-esp32/tree/master/libraries/HTTPUpdate) class and [**HttpsOTAUpdate**](https://github.com/espressif/arduino-esp32/tree/master/libraries/Update/examples/HTTPS_OTA_Update) class.

But the [*ESP32 arduino core documentation*](https://docs.espressif.com/projects/arduino-esp32/en/latest/) is so poor that you probably need to parse the library code provided by the core to know API exactly (in fact, I did it in providing AutoConnect). The [*ESP8266 Documentation Update chapter*](https://arduino-esp8266.readthedocs.io/en/latest/ota_updates/readme.html#http-server) is a clue, but the source code eventually replaces the documentation for it to work properly.

When building an update server over HTTPS, [*HTTPS_OTA_Update.ino*](https://github.com/espressif/arduino-esp32/blob/master/libraries/Update/examples/HTTPS_OTA_Update/HTTPS_OTA_Update.ino) is a clue to know the API. But this example requires embedding the server certificate in the sketch. (which is natural because it's HTTPS)

Another clue is [*httpUpdate.ino*](https://github.com/espressif/arduino-esp32/blob/master/libraries/HTTPUpdate/examples/httpUpdate/httpUpdate.ino), which is included in the ESP32 arduino core example. This suggests the API you're looking for. But it does not say the specifications. After all, we have to clarify it by investigation. Even more confusing is that the ESP32 HTTPUpdate class has incompatible callback exit specifications for notifying update status between the core version 2.0 earlier and later. If your sketch needs to have a UI to inform the update progress, you have no choice but to extend the Update class and implement your own update class or use the HTTPUpdate class since ESP32 arduino core 2.0.

The following sketch is an example of incorporating the HTTPUpdate class with AutoConnect. (Note that the root page is not implemented. Direct access to `/_ac` is required)

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <HTTPUpdate.h>
#include <AutoConnect.h>

static const char OTA[] PROGMEM = R"(
[
  {
    "title": "Update",
    "uri": "/update",
    "menu": true,
    "element": [
      {
        "name": "chk",
        "type": "ACSubmit",
        "value": "Check Update",
        "uri": "/chk"
      }
    ]
  },
  {
    "title": "Update",
    "uri": "/chk",
    "menu": false,
    "element": [
      {
        "name": "res",
        "type": "ACText"
      }
    ]
  }
]
)";

AutoConnect portal;

bool  updateFlag = false;
const char* updateServerUrl = "www.myupdateserver.com";
const int   updateServerPort = 8000;
const char* updateFirmware = "/firmware.bin";

String onUpdate(AutoConnectAux& aux, PageArgument& args) {
  aux["res"].value = "Updating...";
  updateFlag = true;
  return String();
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();
  if (portal.begin()) {
    portal.load(OTA);
    portal.on("/chk", onUpdate);
  }
}

void loop() {
  if (updateFlag) {

    WiFiClient  client;

    t_httpUpdate_return ret = httpUpdate.update(client, updateServerUrl, updateServerPort, updateFirmware);
    switch (ret) {
    case HTTP_UPDATE_FAILED:
      Serial.printf("HTTP_UPDATE_FAILED Error (%d): %s\n", httpUpdate.getLastError(), httpUpdate.getLastErrorString().c_str());
      break;
    case HTTP_UPDATE_NO_UPDATES:
      Serial.println("HTTP_UPDATE_NO_UPDATES");
      break;
    case HTTP_UPDATE_OK:
      Serial.println("HTTP_UPDATE_OK");
      break;
    }
  }

  updateFlag = false;
  portal.handleClient();
}
```

The above sketch does not meet your needs in the following ways:
- The mechanism to detect a valid update has not implemented.
- The mechanism to notify the update progress status has not implemented.
- ESP32 module is forced to reboot after update.
- The *updateserver.py* included with AutoConnect does not fit this sketch. The simplest static HTTP server will suffice as the server that fits the HTTPUpdate class. You can apply it with [http-server](https://www.npmjs.com/package/http-server), which is a typical [npm](https://www.npmjs.com/) package by [Node.js](https://nodejs.org/). `npx http-server -p 8000`

**Reply by adnan6336** [reply]
@Hieromon
Good Day Sir,
I truly appreciated your response and the time you took to explain my question, you really explained each and every thing very well.
So I have managed to successfully update with your given code with some editing.
I tried the sketch with your updateserver.py and it works and updated my device (Node MCU) with the new code however you mentioned that "The updateserver.py included with AutoConnect does not fit this sketch." so I want to know a little more about why does it works and what's the difference and how can I improve it.
Much appreciated your work. 
Many Thanks.

```
#include <Arduino.h>
#include <ESP8266WiFi.h>
#include <ESP8266WiFiMulti.h>
#include <ESP8266HTTPClient.h>
#include <ESP8266httpUpdate.h>
#include <AutoConnect.h>

static const char OTA[] PROGMEM = R"(
[
  {
    "title": "Update",
    "uri": "/update",
    "menu": true,
    "element": [
      {
        "name": "chk",
        "type": "ACSubmit",
        "value": "Check Update",
        "uri": "/chk"
      }
    ]
  },
  {
    "title": "Update",
    "uri": "/chk",
    "menu": false,
    "element": [
      {
        "name": "res",
        "type": "ACText"
      }
    ]
  }
]
)";

AutoConnect portal;

bool  updateFlag = true;
bool pressCheck = false;
const char* updateServerUrl = "192.168.0.110";
const int   updateServerPort = 8000;
const char* updateFirmware = "/updateFilev1.0.ino.nodemcu.bin";

String onUpdate(AutoConnectAux& aux, PageArgument& args) {
  aux["res"].value = "Updating...";
  updateFlag = true;
  
  return String();
}

void setup() {
  pinMode(14,INPUT_PULLUP);
  Serial.begin(115200);
  delay(1000);
  Serial.begin(115200);
  Serial.println();
  if (portal.begin()) {
    portal.load(OTA);
    portal.on("/chk", onUpdate);
  }
  Serial.println("Hello World");
}

void loop() {
  
  
  if (!pressCheck && !digitalRead(14)) {
    Serial.println("trying to update");
    pressCheck=true;
    WiFiClient  client;

    t_httpUpdate_return ret = ESPhttpUpdate.update(client, updateServerUrl, updateServerPort, updateFirmware);
    switch (ret) {
    case HTTP_UPDATE_FAILED:
      Serial.printf("HTTP_UPDATE_FAILED Error (%d): %s\n", ESPhttpUpdate.getLastError(),ESPhttpUpdate.getLastErrorString().c_str());
      break;
    case HTTP_UPDATE_NO_UPDATES:
      Serial.println("HTTP_UPDATE_NO_UPDATES");
      break;
    case HTTP_UPDATE_OK:
      Serial.println("HTTP_UPDATE_OK");
      break;
    }
  }

  updateFlag = false;
  portal.handleClient();
}
```

**Reply by Hieromon** [reply]
@adnan6336 , An `updateserver.py` expects the HTTPUpdate class as its counterpart HTTP client. In the case of ESP8266, it is the `ESP8266HttpUpdate` class and is exported as an `ESP8266httpUpdate` instance.
Your sketch cites an ESP8266httpUpdate instance, so an updateserver.py can send an `updateFilev1.0.ino.nodemcu.bin` file in response to a [`GET`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.cpp#L198) request issued internally by the [`ESP8266HTTPUpdate::handleUpdate`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.cpp#L152) through an [`ESP8266HTTPUpdate::update`]() function. (Assuming the `.bin` file is correctly placed where an updateserver.py is accessible)

On the other hand, the core of an `updateserver.py` is the [`http.server`](https://docs.python.org/3.8/library/http.server.html#module-http.server) Python module and has a [`do_GET`](https://docs.python.org/3.8/library/http.server.html#http.server.SimpleHTTPRequestHandler.do_GET) request handler to reply with a `GET` requeset from [`ESP8266HTTPUpdate::handleUpdate`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.cpp#L198) as a client. The [*do_GET request handler*](https://github.com/Hieromon/AutoConnect/blob/aa86edab5c86b37d6cc2f6c0d0e64b44735b6252/src/updateserver/python3/updateserver.py#L37) in *updateserver.py* does more than just send the requested static content as a `.bin` updater. It returns a list of staged `.bin` updater files in response to the [`list`](https://hieromon.github.io/AutoConnect/otaserver.html#1-http-url-query-for-the-catalog-list-of-the-updatable) request that arrives at `/_catalog`. Then [parses the HTTP request headers](https://github.com/Hieromon/AutoConnect/blob/aa86edab5c86b37d6cc2f6c0d0e64b44735b6252/src/updateserver/python3/updateserver.py#L64) to verify that the update request is from the ESP8266HTTPUpdater.

> it works and updated my device (Node MCU) with the new code however you mentioned that "The updateserver.py included with AutoConnect does not fit this sketch."

It is due to the following reasons; The ESP8266HTTPUpdate::update function can have a [`currentVersion`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.h#L119) argument. So, the sketch can request a newer version of the `.bin` updater than the specified value. In other words, it requires the update server to validate the updater version and then send out a newer updater.
The value of the `currentVersion` argument would be included in the HTTP header as [`x-ESP8266-version`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.cpp#L178) of the [`GET`](https://github.com/esp8266/Arduino/blob/f8de3fb464a220452a31680b7510b2f2420fd656/libraries/ESP8266httpUpdate/src/ESP8266httpUpdate.cpp#L198) request issued internally by the ESP8266HTTPUpdate::handleUpdate request mentioned above. But, The updateserver.py [does not validate](https://github.com/Hieromon/AutoConnect/blob/aa86edab5c86b37d6cc2f6c0d0e64b44735b6252/src/updateserver/python3/updateserver.py#L65) this value and sends it as is without version verification of the requested `.bin` updater.
That is, *updateserver.py* does **not meet the server feature implementation required by the ESP8266HTTPupdate** class and does not conform to it. You can see this behavior by tracing the code in *updateserver.py*, ESP8266HTTPUpdate class and I think you could do it.

Also, I said, "The simplest static HTTP server is enough for a server that conforms to the HTTPUpdate class". AutoConnectUpdate class leaves the version confirmation to the user by using the UI including the Date. So, you don't even need to use `updateserver.py` if it's an update server with no [`/_catalog`](https://hieromon.github.io/AutoConnect/otaserver.html#1-http-url-query-for-the-catalog-list-of-the-updatable) and no version verification.
